text/0018-bee-bundle/0018-bee-bundle.md

```diff
@@ -0,0 +1,281 @@
+
+### Sign
+
+*Client side operation.*
+
+Signing a bundle allow you to prove that you are the owner of the address you are trying to move funds from. With no signature or a bad signature, the bundle won't be considered valid and the funds won't be moved. Only the owner of the right seed is able to generate the right signature for this address.
+
+Pseudocode:
+
+```
+sign(bundle, seed, inputs)
+| current_index = 0
+|
+| for transaction in bundle
+| | if transaction.value < 0
+| | | if transaction.current_index >= current_index
+| | | | input = inputs[transaction.address]
+| | | | fragments = sign(seed, input.index, input.security, transaction.bundle)
+| | | | for fragment in fragments
+| | | | | bundle.at(current_index).signature = fragment
+| | | | | current_index = current_index + 1
+| | else
+| | | current_index = current_index + 1
+```
+
+*Since signature size depends on the security level, a single signature can spread out to up to 3 transactions. `inputs` is an object that contain all unused addresses of a seed with a sufficient balance.*
+
+### Pow
+
+*Client side operation.*
+
+Proof of Work (PoW) allows your transactions to be accepted by the network. On the IOTA network, PoW is only a rate control mechanism. Doing PoW on a bundle means doing PoW on each of its transactions and set trunks and branch accordingly. After PoW, a bundle is ready to be sent to the network.
+
+Pseudocode:
+
+```
+pow(bundle, trunk, branch, mwm)
+| for transaction in rev(bundle)
+| | transaction.trunk = trunk
+| | transaction.branch = branch
+| | if transaction.current_index == transaction.last_index
+| | | branch = trunk
+| | transaction.attachment_timestamp = timestamp()
+| | transaction.attachment_timestamp_lower = 0
+| | transaction.attachment_timestamp_upper = 3812798742493
+| | if transaction.tag.empty()
+| | | transaction.tag = transaction.obsolete_tag
+| | trunk = transaction.pow(mwm)
+
+```
+
+### Validate
+
+*Client side and server side operation.*
+
+Validating a bundle means checking the syntactic and semantic integrity of a bundle as a whole and of its constituent transactions. As bundles are atomic transfers, either all or none of the transactions will be accepted by the network. After validation, transactions of a bundle are candidates to be included to the ledger.
+
+For a bundle to be considered valid, the following assertions must be true:
+
+- bundle has announced size;
+- transactions share the same bundle hash;
+- transactions absolute value doesn't exceed total IOTA supply;
+- bundle absolute sum never exceeds total IOTA supply;
+- order of transactions in the bundle is the same as announced by `current_index` and `last_index`;
+- value transactions have an address ending in `0` i.e. has been generated with Kerl;
+- bundle inputs and outputs are balanced i.e. the bundle sum equals `0`;
+- announced bundle hash matches the computed bundle hash;
+- for spending transactions, the signature is valid;
+
+Pseudocode:
+
+```
+validate(bundle):
```

Comment:
Because a `Bundle` should be a valid object, the validation step should be part of the `build` method.

Reply:
`build` will definitely call `validate`. But I assumed `validate` is a function sufficiently interesting be able to be called on its own. I can imagine situation when we want to validate but doesn't care of the resulting bundle.